crates/autopilot/src/infra/solvers/authenticator.rs

```diff
@@ -0,0 +1,107 @@
+use {
+    crate::{
+        domain::{self, eth},
+        infra::blockchain::{contracts::deployment_address, ChainId},
+    },
+    ethcontract::dyns::DynWeb3,
+    primitive_types::H160,
+};
+
+#[allow(dead_code)]
+#[derive(Debug, Clone)]
+pub struct Authenticator {
+    /// The authenticator contract used for allow-listing solvers to settle.
+    authenticator: contracts::GPv2AllowListAuthentication,
+    /// The safe module that is used to provide special role to EOA.
+    authenticator_role: contracts::Roles,
+    /// The EOA that is allowed to add/remove solvers.
+    authenticator_eoa: ethcontract::Account,
+}
+
+pub struct Addresses {
+    pub settlement: Option<eth::Address>,
+    pub authenticator_eoa: eth::H256,
+}
+
+impl Authenticator {
+    /// Creates an authenticator which can remove solvers from the allow-list
+    pub async fn new(web3: DynWeb3, chain: ChainId, addresses: Addresses) -> Self {
+        let address_for = |contract: &ethcontract::Contract, address: Option<H160>| {
+            address
+                .or_else(|| deployment_address(contract, &chain))
+                .unwrap()
+        };
+
+        let settlement = contracts::GPv2Settlement::at(
```

Comment:
IMO the decision to use a custom address or a deployed address should be done at single point in the backend. For example it doesn't make a lot of sense to have 1 component use one address for the settlement contract and another component another address. Overall I think whoever instantiates the `Authenticator` should be responsible for picking the correct address.

Also the settlement contract doesn't really matter here. We only need the authenticator contract.

Reply:
I suggested this to @Mateo-mro so that `Authenticator` relies less on intermediary objects built between the autopilot startup and `Authenticator` constructor being called -> to be more independent of the rest of the code. 

> For example it doesn't make a lot of sense to have 1 component use one address for the settlement contract and another component another address.

I believe this is also irrelevant for the `Authenticator` as a component that we envision running as a separate process etc. Choosing the right address would be the responsibility of `Authenticator`.